### DISCUSSION TITLE
IModelReadRpcInterface.openForRead() is not available

### DISCUSSION THREAD

**Discussion by ec-dev-1** [iTwin/itwinjs-core#3175] [Q&A] [answered]
Hello there,

I am facing an issue with the iTwin once in production and thought that maybe if I'd share the issue here, I would have more insight while de-bugging.  The issue is that when users log-in, the first log-in appears to be giving this error.  All subsequent logins seem to be fine.  
I do not know much about the IModelReadRpcInterface and reading the Docs didn't help me much. 
What could be possibly be causing this issue?

I do want to add that during development, this issue never happens.

Any insight would be much appreciated.

Thank you!

![Screenshot 2022-02-04 164020](https://user-images.githubusercontent.com/86491141/152608857-5901d0ac-f25f-49eb-b20f-da0a0a04679a.png)

**Comment by aruniverse** [comment]
Hi, 

Could you share the version of iTwin.js and itwin/web-viewer-react you are using?
Could you also look at the network tab, look at the failing request, and share the "X-Correlation-Id" header?

**Reply by ec-dev-1** [reply]
"@bentley/itwin-client": "^2.19.15",
    "@itwin/web-viewer-react": "^1.3.0",


There seems to be no failing request related to Auth or Logging in.
The only failing requests are the ones that I have posted here which might not be related to this issue:
https://github.com/iTwin/itwinjs-core/discussions/3176

Thank you for your prompt reply

**Reply by calebmshafer** [reply]
@ec-dev-1 could you supply the `x-correlation-id` header from the failed openForRead request in your network tab?

**Reply by ec-dev-1** [reply]
I've gone through the network tab and there seems to be no results when I searched up "openForRead".
Would you be able to break it down a bit more?

**Reply by ec-dev-1** [reply]
Update:

I went on development and found what you were talking about.
This request doesn't seem to happen for me in production when the bug happens.

However, once I refresh the page, I am able to see it.

Here's what I get for when it works:  x-correlation-id: a0627646-9ea7-42fb-b9d5-4b15871474e6

**Reply by ec-dev-1** [reply]
Update 2:

On the initial login, this message seems to be appearing on the console.  Thought it might aid in giving insight on what might be happening.
![Screenshot 2022-02-07 101234](https://user-images.githubusercontent.com/86491141/152817315-6d5b1be5-9ca8-496b-ae21-72eced20299c.png)

**Reply by aruniverse** [reply]
Hey, would you be able to reproduce the issue and give us another correlation id? 
That log should be fine.

**Reply by ec-dev-1** [reply]
@aruniverse 
I was able to reproduce the issue by signing out and logging back in.
Once again, when the bug happens, I do not get any request for IModelReadRpcInterface.

However, once I refresh, it works and this is the correlation id that it gave me: 
 x-correlation-id: d5ada8fe-2765-4f87-a870-0e0ed48b66d7

**Reply by aruniverse** [reply]
Hey, we didn't see anything wrong on our servers. Are you guys doing anything in your viewer in the `onIModelAppInit` callback prop? If so, mind sharing a snippet of the code?

**Reply by ec-dev-1** [reply]
@aruniverse   When I searched _onImodelAppInit_ on VS Code, it only gave a result from the bundled JS file.  I do not think there was a direct modification for this on our end.

**Reply by aruniverse** [reply]
Is this imodel specific? Do you run into the same issue with other imodels? Does anyone else on your team run into this?

**Reply by ec-dev-1** [reply]
@aruniverse  
We are currently only using one imodel. Is there a Test IMJS_IMODEL_ID and IMJS_CONTEXT_ID that we can use to verify if the same issue happens with other models?
And yes, other users in the team are able to replicate the issue.
I am also trying to revert to previous builds just in case the issue is on our end.

**Reply by aruniverse** [reply]
You can try creating a new imodel here: https://developer.bentley.com/my-imodels/

**Reply by ec-dev-1** [reply]
@aruniverse 
I've tried creating a new imodel (house) and using that to run the iTwin platform.
Unfortunately, I still faced the error message upon initial log in. 
Refreshing fixes the issue and I can view the model.

This issue was discovered a few weeks ago and so this afternoon I tried using different production builds from early January 2022 and signing-out then logging back in appears to still trigger the issue.  I never caught this bug as I am always logged in on all of my accounts.

Just wanted to share my results as we continue to de-bug what could be causing this.


Thank you

**Comment by ec-dev-1** [comment, marked as answer]
@calebmshafer  @aruniverse  the issue was on our end and we were able to fix it.   We were declaring the bundled JS files in both the Head and Body of the page which greatly affects the timing of how the code runs.  Eliminating it from the Head fixed the issue.

I really appreciate all your assistance.

Thank you

**Reply by aruniverse** [reply]
Glad to hear that's sorted out!
